File: main/ratelimit.c
#include "ratelimit.h"
#include "config.h"
#include "memory.h"
#include "nvs_keys.h"
#include "esp_log.h"
#include <string.h>
#include <stdio.h>
#include <time.h>

static const char *TAG = "ratelimit";

// Simple sliding window using hour/day counters
static int s_requests_this_hour = 0;
static int s_requests_today = 0;
static int s_last_hour = -1;
static int s_last_day = -1;

void ratelimit_init(void)
{
    // Load persisted daily count
    char buf[16];
    if (memory_get(NVS_KEY_RL_DAILY, buf, sizeof(buf))) {
        s_requests_today = atoi(buf);
    }
    if (memory_get(NVS_KEY_RL_DAY, buf, sizeof(buf))) {
        s_last_day = atoi(buf);
    }

    ESP_LOGI(TAG, "Rate limiter initialized: %d requests today", s_requests_today);
}

static void update_time_window(void)
{
    time_t now;
    struct tm timeinfo;
    time(&now);
    localtime_r(&now, &timeinfo);

    int current_hour = timeinfo.tm_hour;
    int current_day = timeinfo.tm_yday;

    // Reset hourly counter if hour changed
    if (current_hour != s_last_hour) {
        s_requests_this_hour = 0;
        s_last_hour = current_hour;
    }

    // Reset daily counter if day changed
    if (current_day != s_last_day) {
        s_requests_today = 0;
        s_last_day = current_day;

        // Persist the new day
        char buf[16];
        snprintf(buf, sizeof(buf), "%d", current_day);
        memory_set(NVS_KEY_RL_DAY, buf);
        memory_set(NVS_KEY_RL_DAILY, "0");

        ESP_LOGI(TAG, "Daily rate limit reset");
    }
}

bool ratelimit_check(char *reason, size_t reason_len)
{
#if !RATELIMIT_ENABLED
    return true;
#endif

    update_time_window();

    // Check hourly limit
    if (s_requests_this_hour >= RATELIMIT_MAX_PER_HOUR) {
        snprintf(reason, reason_len,
                 "Rate limited: %d/%d requests this hour. Try again later.",
                 s_requests_this_hour, RATELIMIT_MAX_PER_HOUR);
        ESP_LOGW(TAG, "Hourly rate limit exceeded");
        return false;
    }

    // Check daily limit
    if (s_requests_today >= RATELIMIT_MAX_PER_DAY) {
        snprintf(reason, reason_len,
                 "Daily limit reached: %d/%d requests today. Resets at midnight.",
                 s_requests_today, RATELIMIT_MAX_PER_DAY);
        ESP_LOGW(TAG, "Daily rate limit exceeded");
        return false;
    }

    return true;
}

void ratelimit_record_request(void)
{
    update_time_window();

    s_requests_this_hour++;
    s_requests_today++;

    // Persist daily count periodically
    char buf[16];
    snprintf(buf, sizeof(buf), "%d", s_requests_today);
    memory_set(NVS_KEY_RL_DAILY, buf);

    ESP_LOGD(TAG, "Request recorded: %d/hour, %d/day",
             s_requests_this_hour, s_requests_today);
}

int ratelimit_get_requests_today(void)
{
    return s_requests_today;
}

int ratelimit_get_requests_this_hour(void)
{
    return s_requests_this_hour;
}

void ratelimit_reset_daily(void)
{
    s_requests_today = 0;
    s_requests_this_hour = 0;
    memory_set(NVS_KEY_RL_DAILY, "0");
    ESP_LOGI(TAG, "Rate limits manually reset");
}

File: src/ratelimit.zig
//! Request rate limiting.
//! Sliding window counters per hour and per day.
//! Replaces ratelimit.c (ESP-IDF NVS calls replaced with file-based memory).

const std = @import("std");
const config = @import("config.zig");
const memory = @import("memory.zig");

const log = std.log.scoped(.ratelimit);

var g_mutex: std.Thread.Mutex = .{};
var g_requests_this_hour: u32 = 0;
var g_requests_today: u32 = 0;
var g_last_hour: i32 = -1;
var g_last_day: i32 = -1;
var g_last_year: i32 = -1;
var g_allocator: std.mem.Allocator = undefined;

pub fn init(allocator: std.mem.Allocator) void {
    g_allocator = allocator;

    // Load persisted daily count
    if (memory.memGet(allocator, config.MEMORY_NAMESPACE, config.KEY_RL_DAILY) catch null) |v| {
        defer allocator.free(v);
        g_requests_today = std.fmt.parseInt(u32, v, 10) catch 0;
    }
    if (memory.memGet(allocator, config.MEMORY_NAMESPACE, config.KEY_RL_DAY) catch null) |v| {
        defer allocator.free(v);
        g_last_day = std.fmt.parseInt(i32, v, 10) catch -1;
    }
    if (memory.memGet(allocator, config.MEMORY_NAMESPACE, config.KEY_RL_YEAR) catch null) |v| {
        defer allocator.free(v);
        g_last_year = std.fmt.parseInt(i32, v, 10) catch -1;
    }

    log.info("Rate limiter initialized: {} requests today", .{g_requests_today});
}

fn updateTimeWindow() void {
    const ts = std.time.timestamp();
    const epoch_seconds = std.time.epoch.EpochSeconds{ .secs = @intCast(ts) };
    const day_seconds = epoch_seconds.getDaySeconds();
    const epoch_day = epoch_seconds.getEpochDay();
    const year_day = epoch_day.calculateYearDay();

    const current_hour: i32 = @intCast(day_seconds.getHoursIntoDay());
    const current_day: i32 = @intCast(year_day.day);
    const current_year: i32 = @intCast(year_day.year);

    if (current_hour != g_last_hour) {
        g_requests_this_hour = 0;
        g_last_hour = current_hour;
    }

    if (current_day != g_last_day or current_year != g_last_year) {
        g_requests_today = 0;
        g_last_day = current_day;
        g_last_year = current_year;

        // Persist
        var buf: [32]u8 = undefined;
        const day_str = std.fmt.bufPrint(&buf, "{}", .{current_day}) catch "0";
        memory.memSet(config.MEMORY_NAMESPACE, config.KEY_RL_DAY, day_str) catch {};
        const year_str = std.fmt.bufPrint(&buf, "{}", .{current_year}) catch "0";
        memory.memSet(config.MEMORY_NAMESPACE, config.KEY_RL_YEAR, year_str) catch {};
        memory.memSet(config.MEMORY_NAMESPACE, config.KEY_RL_DAILY, "0") catch {};

        log.info("Daily rate limit reset", .{});
    }
}

/// Check if a request is allowed. Returns null if allowed, error message if not.
pub fn check(buf: []u8) ?[]const u8 {
    if (!config.RATELIMIT_ENABLED) return null;

    g_mutex.lock();
    defer g_mutex.unlock();

    updateTimeWindow();

    if (g_requests_this_hour >= config.RATELIMIT_MAX_PER_HOUR) {
        return std.fmt.bufPrint(buf,
            "Rate limited: {}/{} requests this hour. Try again later.",
            .{ g_requests_this_hour, config.RATELIMIT_MAX_PER_HOUR }) catch "Rate limited: hourly limit reached.";
    }

    if (g_requests_today >= config.RATELIMIT_MAX_PER_DAY) {
        return std.fmt.bufPrint(buf,
            "Daily limit reached: {}/{} requests today. Resets at midnight.",
            .{ g_requests_today, config.RATELIMIT_MAX_PER_DAY }) catch "Rate limited: daily limit reached.";
    }

    return null;
}

pub fn recordRequest() void {
    g_mutex.lock();
    defer g_mutex.unlock();

    updateTimeWindow();

    g_requests_this_hour += 1;
    g_requests_today += 1;

    // Persist daily count
    var buf: [32]u8 = undefined;
    const daily_str = std.fmt.bufPrint(&buf, "{}", .{g_requests_today}) catch "0";
    memory.memSet(config.MEMORY_NAMESPACE, config.KEY_RL_DAILY, daily_str) catch {};
}

pub fn getRequestsToday() u32 {
    g_mutex.lock();
    defer g_mutex.unlock();
    return g_requests_today;
}

pub fn getRequestsThisHour() u32 {
    g_mutex.lock();
    defer g_mutex.unlock();
    return g_requests_this_hour;
}

pub fn resetDaily() void {
    g_mutex.lock();
    defer g_mutex.unlock();
    g_requests_today = 0;
    g_requests_this_hour = 0;
    memory.memSet(config.MEMORY_NAMESPACE, config.KEY_RL_DAILY, "0") catch {};
    log.info("Rate limits manually reset", .{});
}
